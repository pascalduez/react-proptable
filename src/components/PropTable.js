// @flow
/* eslint-disable no-unused-expressions */
import React from 'react';
import Table from './Table';
import getPropTypes from '../utils';
import type { ReactComponent, Properties } from '../types';

type Props = {
  title?: string,
  component: ReactComponent,
  className?: string,
};


export default function PropTable({ title, component, className, ...extraProps }: Props) {
  const heading = title ||
    [component.displayName, 'PropTypes']
      .filter(Boolean)
      .join(' ');

  const properties: Properties = Object.values(getPropTypes(component));
  const hasProps = properties.length > 0;

  return (
    <section className={className} {...extraProps}>
      {heading &&
        <h1>{heading}</h1>}
      {do {
        if (hasProps) {
          <Table properties={properties} />;
        } else {
          <p>Could not find any propType or type to document.</p>;
        }
      }}
    </section>
  );
}
