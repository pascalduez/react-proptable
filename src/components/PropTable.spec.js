import React, { PropTypes } from 'react';
import { render } from 'enzyme';

import PropTable from './PropTable';

// Fixtures.
const WithoutProps = () => <span>test</span>;
const WithProps = ({ label, foo }) => <span>{`${label}${foo}`}</span>;
WithProps.displayName = 'WithProps';
WithProps.propTypes = {
  label: PropTypes.string.isRequired,
  foo: PropTypes.number,
};
WithProps.defaultProps = {
  foo: '',
};


describe('<PropTable>', () => {
  describe('When there is no propType found', () => {
    it('should render the default markup and classes', () => {
      const wrapper = render(
        <PropTable component={WithoutProps} />
      );

      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('When propTypes are found', () => {
    it('should render the default markup and classes', () => {
      const wrapper = render(
        <PropTable component={WithProps} />
      );

      expect(wrapper).toMatchSnapshot();
    });
  });
});
