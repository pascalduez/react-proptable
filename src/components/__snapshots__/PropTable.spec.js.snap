// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PropTable> When propTypes are found should render the default markup and classes 1`] = `
<section>
  <h1>
    WithProps PropTypes
  </h1>
  <table>
    <thead>
      <tr>
        <th>
          Name
        </th>
        <th>
          Type
        </th>
        <th>
          Required
        </th>
        <th>
          Default
        </th>
        <th>
          Description
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          label
        </td>
        <td>
          <pre>
            <code>
              string
            </code>
          </pre>
        </td>
        <td>
          yes
        </td>
        <td>
          <pre>
            <code>
              -
            </code>
          </pre>
        </td>
        <td>
          -
        </td>
      </tr>
      <tr>
        <td>
          foo
        </td>
        <td>
          <pre>
            <code>
              number
            </code>
          </pre>
        </td>
        <td>
          no
        </td>
        <td>
          <pre>
            <code />
          </pre>
        </td>
        <td>
          -
        </td>
      </tr>
    </tbody>
  </table>
</section>
`;

exports[`<PropTable> When there is no propType found should render the default markup and classes 1`] = `
<section>
  <h1>
    PropTypes
  </h1>
  <p>
    Could not find any propType or type to document.
  </p>
</section>
`;
